### PR TITLE
SslStream - Making the reset buffer logic clearer

### PR DESCRIPTION
**Pull request by Drawaes**
These are to make the "returning" of the buffer logic clearer. Also the asserts caught the fact that the logic was a little odd. We don't need to resize the buffer anymore therefore we don't have to call the "ensureinternalbuffersize" on each read (for the header and the frame). 

So I have removed that, renamed to "ResetReadBuffer". Removed one of the extra checks on the if statement for "if buffercount > 0 && bufferoffset >0". So this has the buffercount removed and an assert added to make sure we only ever get here if the buffer is null or if it has data (if it doesn't have data it should be released).

Finally I added an extra assert to double check there are no decrypted bytes left.

Using @geoffkizer benchmark on Azure windows 4 cores I see no difference in the 256 byte test a slight uptick in the bigger buffer but nothing to write home about.

### CODE REVIEW

**Comment by Drawaes** [comment]
Word of warning on my test results, the 256 byte one seems to jump around from restart to restart on Azure also the time of day. The 4096 and bigger seem to be a lot more stable so I don't really trust much under 2k anyway.

**Comment by geoffkizer** [comment]
LGTM

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 1]
```diff
@@ -204,13 +204,16 @@ internal void EndWrite(IAsyncResult asyncResult)
...
         {
+            Debug.Assert(_decryptedBytesCount == 0);
+            Debug.Assert(_internalBuffer == null || _internalBufferCount > 0);
+
             if (_internalBuffer == null)
             {
                 _internalBuffer = ArrayPool<byte>.Shared.Rent(ReadBufferSize);
             }
-            else if (_internalOffset > 0 && _internalBufferCount > 0)
+            else if (_internalOffset > 0)
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Should the `_internalOffset = 0;` below be moved into this `else` branch?

Comment (Drawaes) [reply src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Yes now it's only needed there, I have moved it

**Pull request merged**